HttpClient retry may cause "unexpected message type: DefaultFullHttpRequest, state: 1"

When using HttpClient for POST requests, connection resets by HTTP server may cause the client to lock in ``unexpected message type: DefaultFullHttpRequest, state: 1`` state for subsequent requests with the same server. It's caused by the retry feature of HttpClient.

## Expected Behavior
Subsequent HTTP requests work normally after a connection reset by the HTTP server.

## Actual Behavior
Sometimes, the HttpClient is locked in ``unexpected message type: DefaultFullHttpRequest, state: 1`` state.
## Steps to Reproduce
The issue has happened in our production when the network was unstable, i.e. with lots of TCP connection resets.

1. Start a HTTP server (could be any HTTP server, it doesn't matter)
2. Using [HttpClient](https://github.com/reactor/reactor-netty/blob/main/reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClient.java) (it's easier to reproduce the issue if setting maxConnections to 1) to send a POST request (with body) to the HTTP server. Debug to pause at [reactor.netty.http.HttpOperations#send](https://github.com/reactor/reactor-netty/blob/ea64a1c5ca74fbb59af272f184b0b031f399ff5e/reactor-netty-http/src/main/java/reactor/netty/http/HttpOperations.java#L113) - ``return new PostHeadersNettyOutbound``
```java
	public NettyOutbound send(Publisher<? extends ByteBuf> source) {
		if (!channel().isActive()) {
			return then(Mono.error(AbortedException.beforeSend()));
		}
		if (source instanceof Mono) {
                         // pause here. Channel must be active in order to reach here
			return new PostHeadersNettyOutbound(((Mono<ByteBuf>)source)
					....
		}
		....
	}
```
3. Kill the HTTP server
4. The HTTP request continues to do ``HttpObjectEncoder#encode``, which will release the refCnt of ``DefaultFullHttpRequest#content`` [here](https://github.com/netty/netty/blob/574e079381f9b947fad979a88fa0b27f306f627d/codec/src/main/java/io/netty/handler/codec/MessageToMessageEncoder.java#L91)
5. Connection reset by HTTP server (at step 3) leads to [reactor.netty.http.client.HttpClientConnect.HttpObserver#onUncaughtException](https://github.com/reactor/reactor-netty/blob/ea64a1c5ca74fbb59af272f184b0b031f399ff5e/reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java#L345), which leads to a retry of the HTTP request
6. Start the HTTP server again
7. Retrying the HTTP request will go through ``HttpObjectEncoder#encode`` again. However, the refCnt of ``DefaultFullHttpRequest#content`` was already released at step 4. Trying to encode again sometimes (not always. it depends on exact moment of the connection reset) causes ```java.lang.IndexOutOfBoundsException``` to be thrown at [here]( https://github.com/netty/netty/blob/574e079381f9b947fad979a88fa0b27f306f627d/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectEncoder.java#L131), which leaves ``HttpObjectEncoder#state`` in ``ST_CONTENT_NON_CHUNK`` state. ``io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1`` is thrown after at [here](https://github.com/netty/netty/blob/574e079381f9b947fad979a88fa0b27f306f627d/codec/src/main/java/io/netty/handler/codec/MessageToMessageEncoder.java#L91).
8. Subsequent requests with the same server (assuming using the same TCP connection) will always meet with error  ``unexpected message type: DefaultFullHttpRequest, state: 1`` because the instance of``HttpObjectEncoder`` was locked in ``ST_CONTENT_NON_CHUNK`` state at step 7.

Stack trace when the above situation happens, that is ``io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1`` and then ``unexpected message type: DefaultFullHttpRequest, state: 1``.
```
2021-04-29 16:55:17.008  WARN   --- [reactor-http-nio-3] r.netty.http.client.HttpClientConnect    : [id: 0xe1894edd, L:/127.0.0.1:65440 - R:/127.0.0.1:11001] The connection observed an error
io.netty.handler.codec.EncoderException: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
  at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
  at io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
  at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:715)
  at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:762)
  at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:788)
  at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:756)
  at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:806)
  at io.netty.channel.DefaultChannelPipeline.writeAndFlush(DefaultChannelPipeline.java:1025)
  at io.netty.channel.AbstractChannel.writeAndFlush(AbstractChannel.java:294)
  at reactor.netty.http.HttpOperations.lambda$send$0(HttpOperations.java:118)
  at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:151)
  at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53)
  at reactor.core.publisher.Mono.subscribe(Mono.java:4090)
  at reactor.netty.NettyOutbound.subscribe(NettyOutbound.java:329)
  at reactor.core.publisher.MonoSource.subscribe(MonoSource.java:58)
  at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
  at reactor.netty.http.client.HttpClientConnect$HttpIOHandlerObserver.onStateChange(HttpClientConnect.java:430)
  at reactor.netty.ReactorNetty$CompositeConnectionObserver.onStateChange(ReactorNetty.java:503)
  at reactor.netty.resources.PooledConnectionProvider$DisposableAcquire.run(PooledConnectionProvider.java:571)
  at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
  at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
  at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
  at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
  at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
  at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  at java.lang.Thread.run(Thread.java:745)
Caused by: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
  at io.netty.util.internal.ReferenceCountUpdater.toLiveRealRefCnt(ReferenceCountUpdater.java:74)
  at io.netty.util.internal.ReferenceCountUpdater.release(ReferenceCountUpdater.java:138)
  at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:100)
  at io.netty.handler.codec.http.DefaultFullHttpRequest.release(DefaultFullHttpRequest.java:102)
  at io.netty.util.ReferenceCountUtil.release(ReferenceCountUtil.java:88)
  at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:91)
  ... 25 common frames omitted
2021-04-29 16:55:28.282 ERROR   --- [reactor-http-nio-3] r.n.http.server.HttpServerOperations     : [id: 0xdd50934f, L:/0:0:0:0:0:0:0:1:9001 - R:/0:0:0:0:0:0:0:1:65450] Error starting response. Replying error status
io.netty.handler.codec.EncoderException: java.lang.IllegalStateException: unexpected message type: DefaultFullHttpRequest, state: 1
  at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
  at io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
  at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:715)
  at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:762)
  at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:788)
  at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:756)
  at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:806)
  at io.netty.channel.DefaultChannelPipeline.writeAndFlush(DefaultChannelPipeline.java:1025)
  at io.netty.channel.AbstractChannel.writeAndFlush(AbstractChannel.java:294)
  at reactor.netty.http.HttpOperations.lambda$send$0(HttpOperations.java:118)
  at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:151)
  at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53)
  at reactor.core.publisher.Mono.subscribe(Mono.java:4090)
  at reactor.netty.NettyOutbound.subscribe(NettyOutbound.java:329)
  at reactor.core.publisher.MonoSource.subscribe(MonoSource.java:58)
  at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
  at reactor.netty.http.client.HttpClientConnect$HttpIOHandlerObserver.onStateChange(HttpClientConnect.java:430)
  at reactor.netty.ReactorNetty$CompositeConnectionObserver.onStateChange(ReactorNetty.java:503)
  at reactor.netty.resources.PooledConnectionProvider$DisposableAcquire.run(PooledConnectionProvider.java:571)
  at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
  at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
  at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
  at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
  at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
  at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: unexpected message type: DefaultFullHttpRequest, state: 1
  at io.netty.handler.codec.http.HttpObjectEncoder.encode(HttpObjectEncoder.java:86)
  at io.netty.handler.codec.http.HttpClientCodec$Encoder.encode(HttpClientCodec.java:167)
  at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:89)
  ... 25 common frames omitted
```

## Possible Solution
<!--- Not obligatory, but you can suggest a fix/reason for the bug. -->
Disable the retry feature of HttpClient.
```java
HttpClient.create().disableRetry(true)
```
With current implementation, retries of any HTTP request with a body (POST, PUT, etc) in case of connection reset are not safe. This feature should probably be removed (or at least disabled by default).

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in. -->
<!--- Especially, always include the version(s) of Reactor library/libraries you used! -->

* Reactor version(s) used: reactor-netty 0.9.18.RELEASE. reactor-core 3.3.15.RELEASE
* Other relevant libraries versions (eg. `netty`, ...): netty 4.1.60.Final
* JVM version (`java -version`):java version "11" 2018-09-25
* OS and version (eg. `uname -a`):Darwin MacBook-Pro.local 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:07:06 PST 2021; root:xnu-7195.81.3~1/RELEASE_X86_64 x86_64

--- a/src/main/java/reactor/netty/http/client/HttpClientConnect.java
+++ b/src/main/java/reactor/netty/http/client/HttpClientConnect.java
@@ -391,9 +391,17 @@ final class HttpClientConnect extends HttpClient {
 					handler.previousRequestHeaders = ops.requestHeaders;
 				}
 			}
-			else if (AbortedException.isConnectionReset(error) && handler.shouldRetry) {
+			else if (handler.shouldRetry && AbortedException.isConnectionReset(error)) {
 				HttpClientOperations ops = connection.as(HttpClientOperations.class);
 				if (ops != null) {
+					// In some cases the channel close event may be delayed and thus the connection to be
+					// returned to the pool and later the eviction functionality to remote it from the pool.
+					// In some rare cases the connection might be acquired immediately, before the channel close
+					// event and the eviction functionality be able to remove it from the pool, this may lead to I/O
+					// errors.
+					// Mark the connection as non-persistent here so that it never be returned to the pool and leave
+					// the channel close event to invalidate it.
+					ops.markPersistent(false);
 					ops.retrying = true;
 				}
 				if (log.isDebugEnabled()) {
@@ -705,7 +713,7 @@ final class HttpClientConnect extends HttpClient {
 				redirect(re.location);
 				return true;
 			}
-			if (AbortedException.isConnectionReset(throwable) && shouldRetry) {
+			if (shouldRetry && AbortedException.isConnectionReset(throwable)) {
 				shouldRetry = false;
 				redirect(toURI.toString());
 				return true;

